### ISSUE TITLE
Roadmap

### ISSUE THREAD

**Issue by silvenon** [reactjs/react-transition-group#630]
We would like to take this library to the next level, so it fits more nicely into modern React code. For usage in function components, react-transition-group's component API is clearly unnecessary, so we would like to eventually provide a hooks API as well.

First we would like to refactor the test suite with as many use cases for react-transition-group as we can think of, to avoid surprises in the future. Afterwards we would like to simultaneously work on building the hook API and fixing an issue with stale nodes (#623, #469). The stale nodes fix would probably be breaking, so we'll work on hooks in parallel to see if we need to make any more breaking changes to accommodate the hooks API.

This is also an opportunity to go through the rest of the issues to see if there are some echoing concerns that we can address in the following releases. 🤞

🚀

**Comment by xeviltimx** [comment]
Hi, Matija!

I really like the library!) It helped me create such complex animation, which you can watch on my [LinkedIn post](https://www.linkedin.com/posts/tim-mukhin_javascript-js-reactjs-activity-6663115697448636423-9Y91)

Аnd with all its simplicity and performance, I think it deserves more attention than it currently has (correct me if I'm wrong).

When I discovered RTG almost a year ago, it was difficult at first to understand, how can I create animations that I needed at the time. I think that was because of the lack of examples and not so clear explanations of core components. And the landing page looks really simple.

I think I have an idea on how to increase the popularity of the library and speed up it's learning by new users. 
It has a couple of action points:

1. Sell the library to a new user:
     - Update landing page design, make it more professional-looking
     - Create a logo (some examples from me below)
     - Add a simple description of what the library does, when to use it, and in combination with which libraries are best
     - Add comparison with other animation libraries by bundle size, performance, API, etc.
<details>
<summary>Logo examples</summary>

![Frame 1 (1)](https://user-images.githubusercontent.com/8345600/91440299-10dea380-e877-11ea-99d8-077824e72d0d.jpg)

</details>


2. Add many examples from "hello-world" level to advanced level.
3. Add a "library" of predefined animation effects. It'll be user-friendly not to force users to create their own animations, but instead to provide several predefined animations.

I can help with all the points above.
What do you think?

**Comment by silvenon** [comment]
I totally agree that this library that is seen as official deserves far more! Every time someone posted an issue where they misunderstood how to use the library, I committed a change to the docs in hopes to eventually make it clear enough, but apparently it needs something different.

> Update landing page design, make it more professional-looking

Sure, I'm open to suggestions!

> Create a logo (some examples from me below, but it's just a couple of minutes of work, nothing serious)

I like the one on the bottom left the most. 😃 Looking forward to seeing the entire design.

> Add a simple description of what the library does, when to use it, and in combination with which libraries are best

Yep, it's a bit hard to pin down, but I think the most common use case is animating presence of an element in the DOM. I'm very much open to suggestions.

> Add comparison with other animation libraries by bundle size, performance, API, etc.

I'm not sure about comparison because other libraries offer more functionality, I see this library as the minimal React way to make the reflow hack, I'm not sure if there are other libraries doing only that. But bundle size couldn't hurt. 👍

> Add many examples from "hello-world" level to advanced level.

Sure, the more examples the merrier!

> Add a "library" of predefined animation effects. It'll be user-friendly not to force users to create their own animations, but instead to provide several predefined animations.

Can't they do that from examples? (The previous point.) Please explain further.

In any case, thanks for wanting to help, it's much appreciated, I'll be happy to review your code.

**Comment by pxwee5** [comment]
Is this also an opportunity to extend some functionality? 
The SwitchTransition `modes` prop is really cool and I hope we can implement it in `TransitionGroup` as well. 
Main idea is to be able to use modes while the child decides what kind of animation it should be using. 

Otherwise let `SwitchTransition` work like `TransitionGroup` where the child is responsible of the animation.

**Comment by silvenon** [comment]
>    Otherwise let `SwitchTransition` work like `TransitionGroup` where the child is responsible of the animation.
>    This ensures that we have consistent API where if we use a wrapper transition (like `TransitionGroup`), then the child components should be responsible for its own transitions.

Isn't this already the case considering that `SwitchTransition` only takes `mode` and `children`?

**Comment by pxwee5** [comment]
I wish it does, but I got an error when used with 2 children.

**Comment by koba04** [comment]
@silvenon @jquense Is anyone maintaining this repo? I wondered it because the last commit not by dependabot is Jun 4 2020.
https://github.com/reactjs/react-transition-group/commits/master

**Comment by silvenon** [comment]
Not for a long time, as you correctly noticed. I had plans, as described in this issue, but I lost my drive. If you're looking for a quick replacement, it looks like [@headlessui/react](https://github.com/tailwindlabs/headlessui/tree/main/packages/%40headlessui-react) is a good choice.

**Comment by jquense** [comment]
I also have much time for this. tbh I'm pretty disenfranchised with the idea of a good general purpose animation component like RTG or at least with me being a good person to do it. Most of these hooks or affordances really need to be built into React itself, like Vue and Svelte both have. The other viable alternative is pure JS driven libraries like framer motion, which are a bit commitment but work great. I've found that in most cases with one off or simple CSS transitions it's just better to do it manually with a `ref` and some imperative code.

**Comment by koba04** [comment]
@silvenon @jquense 
Thank you for your response!

I agree with you that there are now many options to implement animations in React ecosystems and web-standard ways.
But RTG is still being used in many libraries, so I think it would be nice if this is kept maintained.
Actually, I hope that RTG provides hooks API rather than components, but I think it's more important to fix bugs and update dependencies for ecosystems.
I might not be able to have much time on it, but if anyone won't work on it, I'd like to take over RTG.

**Comment by silvenon** [comment]
Yep, this maintenance fadeout is the worst for such a popular library. I thought I was eventually going to deliver on my promise, but I should have realized that my focus has changed. As far as I'm concerned, please take over. 👍

**Comment by jquense** [comment]
happy to hand it off, RTG is heavily depended on and it's much better if it keeps being maintained.

**Comment by koba04** [comment]
@silvenon @jquense Thank you for your understanding. I'll do my best!

**Comment by ling1726** [comment]
@silvenon I'm working on the Fluent UI library, and we're looking into animations in our library. I also agree that the library could do with modernization especially on a hook based approach for animations. Would you be interested in working together on this?

**Comment by silvenon** [comment]
What could the hook API give you that's more convenient for your goal? Did you try [Transition from Headless UI](https://headlessui.com/react/transition)? It's been meeting all of my needs and it's maintained by Tailwind Labs, who have resources for continued maintenance.

**Comment by davidhouweling** [comment]
hi @silvenon, any thoughts around making `nodeRef` a required field instead of being optional?

**Comment by silvenon** [comment]
Would probably be a good idea as part of v5.

**Comment by jquense** [comment]
Do we have a plan or active maintaince right now. It's absolutely ok if the answer is "no", I'm not exactly involved either. Mostly I just want to know if it'd be ok to PR a few ideas that change the API significantly. 

I've written components and hooks to with RTG as the basis few times now and _think_ I finally have some idea what a modern API might be useful. One that has hooks, and can handle web animations as well as declarative css transforms in a simpler fashion (with influence from framer motion)

**Comment by silvenon** [comment]
I'd be into reviewing these PRs and talking about these ideas. But I wouldn't have the capacity to maintain.
